From 3fc4c079c059937a4d560a9e0b0535c44be8c124 Mon Sep 17 00:00:00 2001
From: Raviteja Narayanam <raviteja.narayanam@xilinx.com>
Date: Tue, 8 Sep 2020 17:15:16 +0530
Subject: [PATCH 1576/1851] i2c: xiic: Add wait for FIFO empty in send_tx

commit 1f764516fb2bec43e218aa336f867188c551c969 from
https://github.com/Xilinx/linux-xlnx.git

If the tx_half_empty interrupt comes first instead of tx_empty,
STOP bit is generated even before all the bytes are transmitted
out on the bus.
STOP bit should be sent only after all the bytes in the FIFO are
transmitted out of the FIFO. So wait until FIFO is empty before sending
the STOP bit.

Signed-off-by: Raviteja Narayanam <raviteja.narayanam@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/i2c/busses/i2c-xiic.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/i2c/busses/i2c-xiic.c b/drivers/i2c/busses/i2c-xiic.c
index 860acf6ff68a..4deed040e05a 100644
--- a/drivers/i2c/busses/i2c-xiic.c
+++ b/drivers/i2c/busses/i2c-xiic.c
@@ -468,6 +468,12 @@ static void xiic_send_tx(struct xiic_i2c *i2c)
 
 		if (i2c->nmsgs == 1) {
 			u8 cr;
+			int status;
+
+			/* Wait till FIFO is empty so STOP is sent last */
+			status = xiic_wait_tx_empty(i2c);
+			if (status)
+				return;
 
 			/* Write to CR to stop */
 			cr = xiic_getreg8(i2c, XIIC_CR_REG_OFFSET);
-- 
2.31.1

